src/lower/lowerer_impl_imperative.cpp

```diff
@@ -3568,12 +3594,32 @@ Stmt LowererImplImperative::codeToIncIteratorVars(Expr coordinate, IndexVar coor
   for (auto& iterator : levelIterators) {
     Expr ivar = iterator.getIteratorVar();
     if (iterator.isUnique()) {
-      Expr increment = iterator.isFull()
-                     ? 1
-                     : ir::Cast::make(Eq::make(iterator.getCoordVar(),
-                                               coordinate),
-                                      ivar.type());
-      result.push_back(compoundAssign(ivar, increment));
+      // BEGIN GJ
+      if (iterator.isFull()) {
+        Expr increment = 1;
+        result.push_back(compoundAssign(ivar, increment));
+      } else {
+        Expr iteratorParentPos = iterator.getParent().getPosVar();
+        ModeFunction iterBounds = iterator.posBounds(iteratorParentPos);
+        result.push_back(iterBounds.compute());
+        auto iteratorVar = iterator.getIteratorVar();
+        vector<Expr> iterBSArgs = {
+          iterator.getMode().getModePack().getArray(1),
+          iteratorVar, iterBounds[1],
+          coordinate,
+        };
+        result.push_back(ir::Assign::make(iteratorVar, ir::Call::make("taco_gallop", iterBSArgs, iteratorVar.type())));
```

Comment:
We have to make sure that the iterator is ordered before we try to use gallop as well.

Reply:
Not sure I understand this; does "ordered" mean "sorted"? I thought the iterator will always be sorted?

Reply:
Yes, ordered means sorted. It's not necessarily the case that an iterator is ordered! For the `Compressed` format used in CSR matrices it is true, but imagine a format based on a hash table. Such a format does not support ordered iteration, but supports fast insertions and lookups.